Civi/Core/Url.php

```diff
@@ -0,0 +1,353 @@
+<?php
+
+namespace Civi\Core;
+
+/**
+ * Generate a URL.
+ *
+ * As input, this class takes a *logical URI*, which may include a range of configurable sub-parts (path, query, fragment, etc).
+ *
+ * As output, it provides a *concrete URL* that can be used by a web-browser to make requests.
+ */
+class Url {
+
+  /**
+   * @var string
+   *   Ex: 'frontend', 'backend'
+   */
+  private $scheme;
+
+  /**
+   * @var string
+   *   Ex: 'civicrm/dashboard'
+   */
+  private $path;
+
+  /**
+   * @var string
+   *   Ex: abc=123&xyz=456
+   */
+  private $query;
+
+  /**
+   * @var string|null
+   */
+  private $fragment;
+
+  /**
+   * Preferred format.
+   *
+   * Note that this is not strictly guaranteed. It may sometimes return absolute URLs even if you
+   * prefer relative URLs (e.g. if there's no easy/correct way to form a relative URL).
+   *
+   * @var string|null
+   *   'relative' or 'absolute'
+   *   NULL means "decide automatically"
+   */
+  private $preferFormat;
+
+  /**
+   * Whether to HTML-encode the output.
+   *
+   * Note: Why does this exist? It's insane, IMHO. There's nothing intrinsically HTML-y about URLs.
+   * However, practically speaking, this class aims to replace `CRM_Utils_System::url()` which
+   * performed HTML encoding by default. Retaining some easy variant of this flag should make the
+   * off-ramp a bit smoother.
+   *
+   * @var bool
+   */
+  private $htmlEscape = FALSE;
```

Comment:
LOL my first thought when I saw the title of this PR was "please can we kill off `$htmlize`"!
Guess not... but at least you called it insane, so if not the death penalty, at least it received a good taunting.